A fix for a browser viewport height calculation

kirlat opened the pull request:
A fix for a viewport height not being calculated correctly when a browser is in a quirks mode. See #205.

balmas commented on packages/components/src/lib/utility/platform.js:
```diff
@@ -32,16 +32,36 @@ export default class Platform {
         ? Math.min(window.innerWidth, document.documentElement.clientWidth, document.body.clientWidth)
         : document.body.clientWidth || window.innerWidth || document.documentElement.clientWidth,
 
+      // The following should calculate a viewport height correctly in both quirks and non-quirks modes
```
but you're still only using documentElement.clientHeight here? why is the clientHeight const only used for the scrollbars?

kirlat replied:
This works, but is not very precise. However, extreme precision was not important for our purpose so I decided not to change it because it is a tested code that is used in several places.

`window.innerHeight` works correctly in both modes. This is a height of a viewport including a scrollbar.

In a "no-quirks" mode `documentElement.clientHeight` is a height of a viewport without a scrollbar. In a "quirks" mode is it a height of a document (a value usually higher than a viewport width). So, depending on the mode and a scrollbar presence `Math.min(window.innerHeight, document.documentElement.clientHeight)` will return a height of a viewport with or without a scrollbar. It is not a problem for the popup height as we have extra margins of 60px and, because scrollbar height is 17px, it will be an extra gap of 17px sometimes, almost not noticeable. It will appear only when a popup is too tall and its height is restricted.

But since I'm touching this probably it will be good to make adjustments here too. Maybe we can have a `viewport`  (or maybe call it outerViewport) to report dimensions including a scrollbar (i.e. innerWidth/innerHeight), then have a scrollbar width props, and then a getters that will return width and height of of a viewport without sidebars (maybe we can call it `innerViewport`? Not sure what's the best name for it)?

What do you think?

balmas replied:
I think that's a good idea. Because I think otherwise we could come back to this at some point and forget why we only applied it in one place. inner and outer viewport sound like ok names to me. thanks.

kirlat replied:
Great, will update it tomorrow.

kirlat commented:
I've added calculations of viewport dimensions with and without scrollbars. I decided to refer to the "outer" dimension as simply "viewport width/height" because that's consistent with a widely accepted practices.
Dimensions without scrollbars are referred to as "inner width/height". I've added comments explaining meanings of those values.

@balmas: please re-review when you will have time. Thanks!

balmas commented on packages/components/src/lib/utility/platform.js:
```diff
@@ -32,14 +32,42 @@ export default class Platform {
-        : window.innerHeight || document.documentElement.clientHeight
+    this.viewport = {
+      // A width of the viewport that includes the width of the scrollbars, if they are shown
+      width: window.innerWidth,
```
is there any chance window.innerWidth is not defined? Is it safe to drop the fallbacks? Same question for innerHeight

kirlat replied:
I did not find any information about cases where it might not behave as expected, except for buggy implementations in some very early versions of some browsers that are now not in use now. I also checked it on several sites known to be problematic, and it worked well there.

I will probably merge it now: it will go through a series of tests anyway, and if there will be any issues, we'll notice and will have time to fix them.

balmas replied:
sounds good.
